# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/RuntimeService.cpp
# Commit: bd7c4e7e5994
# Full Hash: bd7c4e7e59946c7503a98d067ef001a8f5dfc548
# Author: Andrew Sutherland <asutherland@asutherland.org>
# Date: 2020-05-14 21:11:14
# Regressor Bug: 1594572
# File Overlap Count: 1
# Description:
#   Bug 1634259 - Improve worker shutdown edge case handling. r=perry
#   
#   Bug 1594572 attempted to fix a shutdown edge-case where a worker that was
#   started late enough would fail to create a PBackground connection and
#   early return.  This early return, however, left the worker never scheduled
# ==============================================================================

diff -r 80ba3f3cfaf9 -r bd7c4e7e5994 dom/workers/RuntimeService.cpp
--- a/dom/workers/RuntimeService.cpp	Thu May 14 13:42:38 2020 +0000
+++ b/dom/workers/RuntimeService.cpp	Thu May 14 07:44:44 2020 +0000
@@ -2219,8 +2219,9 @@
   auto failureCleanup = MakeScopeExit([&]() {
     // The creation of threadHelper above is the point at which a worker is
     // considered to have run, because the `mPreStartRunnables` are all
-    // re-dispatched after `mThread` is set.
-    mWorkerPrivate->ScheduleDeletion(WorkerPrivate::WorkerRan);
+    // re-dispatched after `mThread` is set.  We need to let the WorkerPrivate
+    // know so it can clean up the various event loops and delete the worker.
+    mWorkerPrivate->RunLoopNeverRan();
   });
 
   mWorkerPrivate->AssertIsOnWorkerThread();